clients/python/girder_client/cli.py

```diff
@@ -49,78 +52,151 @@ def __init__(self, username, password, host=None, port=None, apiRoot=None,
-            raise Exception('localsync command only accepts parent-type of folder')
-
-        gc.loadLocalMetadata(args.local_folder)
-        gc.downloadFolderRecursive(args.parent_id, args.local_folder, sync=True)
-        gc.saveLocalMetadata(args.local_folder)
+class _DeprecatedOption(click.Option):
+    def get_help_record(self, ctx):
+        pass
+
+
+_CONTEXT_SETTINGS = dict(help_option_names=['-h', '--help'])
+
+
+@click.group(context_settings=_CONTEXT_SETTINGS)
+@click.option('--api-url', default=None,
+              help='RESTful API URL '
+                   '(e.g https://girder.example.com:443/%s)' % GirderClient.DEFAULT_API_ROOT)
+@click.option('--api-key', envvar='GIRDER_API_KEY', default=None,
+              help='[default: GIRDER_API_KEY env. variable]')
+@click.option('--username', default=None)
+@click.option('--password', default=None)
+# Deprecated options
+@click.option('--host', default=GirderClient.DEFAULT_HOST, show_default=True,
+              cls=_DeprecatedOption)
+@click.option('--scheme', default=GirderClient.DEFAULT_SCHEME, show_default=True,
+              cls=_DeprecatedOption)
+@click.option('--port', default=GirderClient.DEFAULT_PORT, show_default=True,
+              cls=_DeprecatedOption)
+@click.option('--api-root', default=GirderClient.DEFAULT_API_ROOT,
+              help='relative path to the Girder REST API', show_default=True,
+              cls=_DeprecatedOption)
+@click.pass_context
+def main(ctx, username, password, api_key, api_url, scheme, host, port, api_root):
+    """Perform common Girder CLI operations.
+
+    The CLI is particularly suited to upload (or download) large, nested
+    hierarchy of data to (or from) Girder from (or into) a local directory.
+
+    The recommended way to use credentials is to first generate an api key
+    and then specify the ``api-key`` argument or set the ``GIRDER_API_KEY``
+    environment variable.
+
+    The client also supports ``username`` and ``password`` args. If only the
+    ``username`` is specified, the client will prompt the user to interactively
+    input his/her password.
+    """
+    ctx.obj = GirderCli(
+        username, password, host=host, port=port, apiRoot=api_root,
+        scheme=scheme, apiUrl=api_url, apiKey=api_key)
+
+
+def _lookup_parent_type(client, object_id):
+
+    object_id = client._checkResourcePath(object_id)
+
+    for parent_type in ['folder', 'collection', 'user', 'item']:
+        try:
+            client.get('resource/%s/path' % object_id, parameters={'type': parent_type})
+            return parent_type
+        except HttpError as exc_info:
+            if exc_info.status == 400:
+                continue
+            if sys.version_info[0] >= 3:
+                raise exc_info.with_traceback(sys.exc_info()[2])
+            else:
+                raise (sys.exc_info()[0], sys.exc_info()[1], sys.exc_info()[2])  # noqa: E999
+
+
+def _common_parameters(path_exists=False, path_writable=True,
```

Comment:
For consistency, let's stick to camelCase for the function names in this library.